reactor-core/src/main/java/reactor/core/scheduler/BoundedElasticScheduler.java

```diff
@@ -160,61 +126,112 @@ BoundedScheduledExecutorService createBoundedExecutorService() {
-			if (BOUNDED_SERVICES.compareAndSet(this, services, newServices)) {
-				ScheduledExecutorService e = Executors.newScheduledThreadPool(1, EVICTOR_FACTORY);
-				if (EVICTOR.compareAndSet(this, null, e)) {
-					try {
-						e.scheduleAtFixedRate(newServices::eviction, ttlMillis, ttlMillis, TimeUnit.MILLISECONDS);
-					}
-					catch (RejectedExecutionException ree) {
-						// the executor was most likely shut down in parallel
-						if (!isDisposed()) {
-							throw ree;
-						} // else swallow
-					}
+			if (b == null) {
+				b = SchedulerState.fresh(
+						new BoundedServices(this),
+						Executors.newScheduledThreadPool(1, EVICTOR_FACTORY)
+				);
+			}
+
+			if (STATE.compareAndSet(this, a, b)) {
+				try {
+					b.evictor.scheduleAtFixedRate(b.boundedServices::eviction,
+							ttlMillis, ttlMillis, TimeUnit.MILLISECONDS);
+					return;
 				}
-				else {
-					e.shutdownNow();
+				catch (RejectedExecutionException ree) {
+					// the executor was most likely shut down in parallel
+					if (!isDisposed()) {
+						throw ree;
+					} // else swallow
 				}
-				return;
 			}
 		}
 	}
 
 	@Override
 	public void dispose() {
-		BoundedServices services = BOUNDED_SERVICES.get(this);
-		if (services != SHUTDOWN && BOUNDED_SERVICES.compareAndSet(this, services, SHUTDOWN)) {
-			ScheduledExecutorService e = EVICTOR.getAndSet(this, null);
-			if (e != null) {
-				e.shutdownNow();
+		SchedulerState currentState = STATE.get(this);
+		final BoundedState[] toAwait = currentState.boundedServicesBeforeShutdown.dispose(true);
+
+		SchedulerState previous = STATE.getAndUpdate(this, old -> {
+			if (old.boundedServices != SHUTDOWN && old.boundedServices != SHUTTING_DOWN) {
+				return SchedulerState.terminated(old, toAwait);
 			}
-			services.dispose();
+			return old;
+		});
+
+		if (previous.evictor != null) {
+			previous.evictor.shutdownNow();
 		}
 	}
 
+	@Override
+	public Mono<Void> disposeGracefully(Duration gracePeriod) {
+		return Mono.defer(() -> {
+			for (;;) {
+				SchedulerState currentState = STATE.get(this);
+
+				if (currentState.boundedServices == SHUTTING_DOWN) {
+					// A rival thread is performing the shutdown, keep spinning.
+					continue;
+				}
+
+				if (currentState.boundedServices == SHUTDOWN) {
+					return Mono.whenDelayError(
+							currentState.boundedServicesOnDispose,
+							currentState.evictorOnDispose
+					);
+				}
+
+				SchedulerState terminating = terminating(currentState);
+
+				if (STATE.compareAndSet(this, currentState, terminating)) {
+					// We need an intermediate state to make sure the disposal happens just once
+					// and that we capture the BoundedStates, otherwise we'd lose references to schedulers
+					// that we need to wait for.
+					final BoundedState[] toAwait = terminating.boundedServicesBeforeShutdown.dispose(false);
+					if (terminating.evictor != null) {
+						terminating.evictor.shutdown();
+					}
+
+					SchedulerState terminated = terminated(terminating, toAwait);
```

Comment:
looks like terminated with a reference to an old awaiter is used only here, this my suggestion would be to either, use the current login here only, or creating a separate method. Sounds like for all other scneratios the old references is redundant

Reply:
The `terminated` state of `BoundedElasticScheduler` is used 3 times in this class - twice it uses the initial state passed as arguments to preserve the disposal references and once with null upon start. It might be confusing that there's a `SchedulerState` as a general type in the package, while this implementation is different and makes sense for this particular scheduler.